pkg/test/framework/components/echo/echotest/utils.go

```diff
@@ -0,0 +1,124 @@
+package echotest
+
+import (
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+)
+
+// T ... TODO document me
+type T struct {
+	// rootCtx is the top level test context to generate subtests from and should only be referenced from RunX methods.
+	rootCtx framework.TestContext
+
+	sources      echo.Instances
+	destinations echo.Instances
+
+	sourceDeploymentSetup []func(ctx framework.TestContext, src echo.Instances) error
+	deploymentPairSetup   []func(ctx framework.TestContext, src, dst echo.Instances) error
+}
+
+// New ... TODO document me
+func New(ctx framework.TestContext, instances echo.Instances) *T {
+	s, d := make(echo.Instances, len(instances)), make(echo.Instances, len(instances))
+	copy(s, instances)
+	copy(d, instances)
+	return &T{rootCtx: ctx, sources: s, destinations: d}
+}
+
+type (
+	perDeploymentTest func(ctx framework.TestContext, instances echo.Instances)
+	perInstanceTest   func(ctx framework.TestContext, src echo.Instance, dst echo.Instances)
+)
+
+// Setup runs the given function in the source deployment context. For example, given apps a, b, and c in 2 clusters,
+// these tests would all run before the the context is cleaned up:
+//     a/to_b/from_cluster-1
+//     a/to_b/from_cluster-2
+//     a/to_c/from_cluster-1
+//     a/to_c/from_cluster-2
+//     cleanup...
+//     b/to_a/from_cluster-1
+//     ...
+func (t *T) Setup(setupFn func(ctx framework.TestContext, src echo.Instances) error) *T {
```

Comment:
So any higher level setup (ie regardless of source/dest), should be done outside of echotest?

Reply:
Yes. We can add another for aesthetic purposes but I don't think it's needed.